doc: first cut at documentation for support level

mhdawson opened the pull request.

Emuentes commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,118 @@
+a module can cause friction and stress on both the package maintainer and
+the consumer. This is further complicated because the level of support
+provides for dependencies of a package may be different than that
+targetted by the package a consumer is directly consuming.
```
@mhdawson 
~targetted~
`targeted`
one too many "T"s

mhdawson replied:
@Emuentes, the easiest way for me is for you to use the recommended change feature so that I can just accept them into the doc.

Emuentes commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,118 @@
+consumer and those of the maintainer and reduce the friction and stress
+that can arise due to the gap.
+
+There are 4 dimentions that we standardize in this practice which are:  
```
@mhdawson 
~dimentions~
`dimensions`
"t" should be "s"

Emuentes commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,118 @@
+* backing
+
+For each of these dimentions the potential options are not an ordered list.
+Instead like licences each ID will identify a unique option. If you wish to
```
@mhdawson 
Suggestion:
`Instead, like licenses, each...`

Just added some commas.

Emuentes commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,118 @@
+
+For each of these dimentions the potential options are not an ordered list.
+Instead like licences each ID will identify a unique option. If you wish to
+identify your package with an option which is not let listed, please PR your
```
@mhdawson 
"let" should be "yet"

Emuentes commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,118 @@
+identify your package with an option which is not let listed, please PR your
+option into the lists in the sections which follow.
+
+The recommened baseline practice is to include an entry in the package.json
```
@mhdawson 
~ recommened~
`recommended`
missing a "d" toward the end

Emuentes commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,118 @@
+The support target captures the level of support that the package maintainer
+aims to provide.  The standardized options are as follows:
+
+* ABANDONED - Not recommened for use. The package is depercated or not longer maintained.
```
@mhdawson 
~depercated~
`deprecated`

"no longer maintained" rather than "not longer"

Emuentes commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,118 @@
+aims to provide.  The standardized options are as follows:
+
+* ABANDONED - Not recommened for use. The package is depercated or not longer maintained.
+* NONE - use at your own risk, no active support may or may not work for a given Node.js version.
```
@mhdawson 

suggestion:
comma after "no active support"

Emuentes commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,118 @@
+* NODE-LATEST - The package is maintained only for the Latest Node.js versions. You will be required to update
+  to the latest Node.js version in order to ensure you can use new versions/get security fixes, get updates
+* NODE-LTS - The package is maintained for the active Node.js LTS releases. Anyone creating an application
+  using an LTS version of Node.js and using the latest major version of LTS adopting modules will will not
```
@mhdawson 
"...adopting modules ~will~ will have to accept..."

ljharb commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,118 @@
+a module can cause friction and stress on both the package maintainer and
+the consumer. This is further complicated because the level of support
+provides for dependencies of a package may be different than that
+targetted by the package a consumer is directly consuming.
```
```suggestion
targeted by the package a consumer is directly consuming.
```

Emuentes replied:
Sorry for the duplicate typos I commented on @mhdawson

I didn't see that @ljharb had caught most of the same typos I did!

ljharb commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,118 @@
+consumer and those of the maintainer and reduce the friction and stress
+that can arise due to the gap.
+
+There are 4 dimentions that we standardize in this practice which are:  
```
```suggestion
There are 4 dimensions that we standardize in this practice which are:  
```

ljharb commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,118 @@
+* response-paid
+* backing
+
+For each of these dimentions the potential options are not an ordered list.
```
```suggestion
For each of these dimensions the potential options are not an ordered list.
```

ljharb commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,118 @@
+* backing
+
+For each of these dimentions the potential options are not an ordered list.
+Instead like licences each ID will identify a unique option. If you wish to
```
```suggestion
Instead like licenses each ID will identify a unique option. If you wish to
```

ljharb commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,118 @@
+identify your package with an option which is not let listed, please PR your
+option into the lists in the sections which follow.
+
+The recommened baseline practice is to include an entry in the package.json
```
```suggestion
The recommended baseline practice is to include an entry in `package.json`
```

ljharb commented on docs/drafts/Baseline practive - Document support levels.md:
````diff
@@ -0,0 +1,118 @@
+file for a package which documents the maintainers expectations. For example:
+
+```json
+{ "support": {
````
i do not think we should be encouraging this unless we also provide a tool that can validate its contents.

ljharb replied:
A separate CLI tool would be fine; for example, the SPDX in a license field can be checked with a CLI tool, as well as with npm. The "browser" field, however, has a defacto spec that all tools implicitly validate when accessing it.

Encouraging unchecked usage of a package.json field imo is a really difficult bell to unring.

wesleytodd replied:
IMO the tools I have used with bespoke `package.json` additions have all been fine.

I understand your concern, but of all the source for recommending an addition to the file, a node WG is a good place to do it.  We should obviously vet it with a larger audience first, but if that all goes well I am not sure why this would be a bad recommendation.

ljharb replied:
If any form of schema validation exists that can be used in CI, and we can recommend that as well, then I'd be very happy with this.

mhdawson replied:
A simple node program that loaded package.json and validated the support section should be pretty easy do write I think.  We could land this as "draft" (which is the directory it is targetted to) and look to get a volunteer to write that tool before we move it of draft to the directory for recommended practices.

Emuentes commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,118 @@
+  
+## Support-response and Support-response-paid
+
+Support response quantifies how quickly the maintainer targets/is able to respond to issues:
```
@mhdawson 
Suggestion:
"the maintainer `["chooses to" | "elects to"], or` is able to, respond to issues"

I'm probably throwing too many commas around on this one :-D

Emuentes commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,118 @@
+
+* ABANDONED - Don't expect a response, the package is not being actively maintained
+* BEST-EFFORT - The maintainer is interesed in fixing/discussing issues, however, it's not their day job so there should be 
+  no expectation on response times. If an when the maintainer has time they may respond.
```
@mhdawson 
"If an`d`"

ljharb commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,118 @@
+  anyone creating an application using an LTS version of Node.js and using the latest major version of LTS
+  adopting modules will will not have to accept semver-major level (ie. breaking) changes into that
+  application in order to receive essential fixes.
+* NODE-SUPERSET - The package is maintained for versions of Node.js including both LTS and non-LTS releases, however, it
```
NODE-SUPERSET and NODE-SUPERSET-LTS seem the same except that perhaps the latter is saying "no breaking changes", which is so unrealistic that i don't think any package ever does it. I'm not sure there needs to be a difference here.

mhdawson replied:
Ok, was not sure if anybody would try to achieve that or not.  I'll remove.

Emuentes commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,118 @@
+  a "we read your issue" response. Further work will depend on prioritization of the issue by the maintainer team.
+* REGULAR-1 - There are dedicated resources who regularly maintain the module, expected response time is 1 days or less for
+  a "we read your issue" response. Further work will depend on prioritization of the issue by the maintainer team.
+* 24-7 - There are dedicated resources who regularly maintain the module and they are available 24/7. You can expecct to 
```
@mhdawson 
extra "C" in "expect"

Emuentes commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,118 @@
+  be able to contact the maintainers and get an initial response with 6 hours.
+   
+ Support-response indicates the expectation unless you have paid one of the maintainers or a company that provides
+ 3rd party support. If paid support is available then Supporet-response-paid is an array of one or more options listing the
```
@mhdawson 
~Supporet-response-paid~
`Support-response-paid`

Emuentes commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,118 @@
+ ## Support-backing
+ 
+ * ABANDONED - There is nobody backing this module
+ * HOBBY - The single mainainer maintains the package for fun, does not get any support to continue maintenance.
```
@mhdawson 
missing T in "maintainer"

Eomm commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,118 @@
+* NODE-SUPERSET-LTS - The package is maintained for versions of Node.js including both LTS and non-LTS releases. Documentation
+  for the package will include the non-LTS releases for which the package is still maintained. As with NODE-LTS 
+  anyone creating an application using an LTS version of Node.js and using the latest major version of LTS
+  adopting modules will will not have to accept semver-major level (ie. breaking) changes into that
```
```suggestion
  adopting modules will not have to accept semver-major level (ie. breaking) changes into that
```

mcollina commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,118 @@
+# Baseline practice - Document support support levels
+
+With the larger number of packages within the JavaScript ecosystem the level of
```
I would rewrite this sentences using "Maintaners" as the subject and first word.

mhdawson replied:
@mcollina need a bit more to follow the suggestion.  Can you provide your suggested sentence?

mcollina commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,118 @@
+With the larger number of packages within the JavaScript ecosystem the level of
+support provided by a maintainer varies greatly.  A mismatch between the
+support a maintainer is planning to provide and that expected by a consumer of
+a module can cause friction and stress on both the package maintainer and
```
```suggestion
a module will cause friction and stress on both the package maintainer and
```

mcollina commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,118 @@
+  adopting modules will will not have to accept semver-major level (ie. breaking) changes into that
+  application in order to receive essential fixes.
+* NODE-SUPERSET - The package is maintained for versions of Node.js including both LTS and non-LTS releases, however, it
+  may be necessary to accept semver-major level (ie. breaking) changes into that application in order to receive essential fixes.
```
I would prefer these levels to be organized inside a table with examples on Node LTS releases.

mhdawson replied:
@mcollina I don't quite understand 'with examples on Node LTS releases'?

vweevers commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,118 @@
+{ "support": {
+    "target" : "NODE_LTS",
+    "response": "BEST_EFFORT",
+    "backing": "INDIVIDUAL"
```
`INDIVIDUAL` is not listed as an option below.

mhdawson replied:
Fixed.

dominykas commented on docs/drafts/Baseline practive - Document support levels.md:
````diff
@@ -0,0 +1,118 @@
+```json
+{ "support": {
+    "target" : "NODE_LTS",
+    "response": "BEST_EFFORT",
````
Is there a way to avoid multi-word keywords? There's underscores here and above, but dashes below.

Eg could "Node LTS" be just LTS? Leave the rest to the "engine" field? Or possibly allow multiple entries under "support", for eg platform node, target LTS; platform deno, target latest?

"Best effort" might need another word... (Nothing springs into my mind just now, trying to avoid negative connotations).

But basically anything to avoid "dash" vs "underscore" decisions :)

mhdawson replied:
@wesleytodd it should be any string, and we validate known ones.  

In terms of the starting ones though I'll think about whether we can eliminate dash/underscore or at the very least make them consistent.

ljharb commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,114 @@
+
+Support response quantifies how quickly the maintainer targets/is able to respond to issues:
+
+* ABANDONED - Don't expect a response, the package is not being actively maintained
```
an alternative is "finished" - as many packages are considered to be "done" and need no further attention or updates.

bnb commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,115 @@
+
+* ABANDONED - Not recommened for use. The package is depercated or not longer maintained.
+* NONE - use at your own risk, no active support may or may not work for a given Node.js version.
+* NODE-LATEST - The package is maintained only for the Latest Node.js versions. You will be required to update
```
I would prefer we drop the `NODE-` prefix here (and in all other cases) since it doesn't really add value. Additionally, it prevents community expansion on using this context for the broader system.

mhdawson commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,114 @@
+option into the lists in the sections which follow.
+
+The recommended baseline practice is to include an entry in `package.json`
+file for a package which documents the maintainers expectations. For example:
```
I like that, will add.

mhdawson commented:
Thanks for all the good comments.  Will do a pass to incorporate early next week.

mcollina commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,115 @@
+   maintain the package. Consider supporting this sponsorship through patreon etc.
+ * BOUNTY - The package is maintained through the use of a bounty service
+ * PROJECT - The package is maintained under the auspices of a larger project (ex Node.js project).
+ * FOUNDATION - The package is maintained and supported under the auspices of a Foundation.
```
The Node.js Foundation does not provide financial support to the active development of our project, but it only provides ancillary. However, both BOUNTY and COMPANY provides actual development time.

rxmarbles replied:
I don't think having `FOUNDATION` here is a bad thing. It just implies that those under the foundation provide the most support to the repository right? most of the time those packages are worked on pretty quickly depending on the issues at hand.

mcollina replied:
This section is about money being actively spent on the project. Foundation means that the foundation pays some developers to work on a package. I don’t see how that is different from SPONSORED.

mhdawson replied:
@mcollina I don't think backing is about $. It's about support which can be people time, machine resources or $.  In that context I think being part of a foundation or project does provide some useful context, although that might vary by Foundation.

mcollina replied:
A project that has:

* most developers working on it has an HOBBY.
* some developers supported by their COMPANY

how would be categorized? How that would change if it is part of a Foundation or not?

mhdawson replied:
I would categorize it at COMPANY in your example. The goal is to give some insight into the backing behind the package.  If COMPANY applies then HOBBY is probably a bonus.

In the case of a FOUNDATION it's harder to choose as it depends on the level of COMPANY support and at what level the package is a member of a foundation (for example IMPACT versus Sandbox).  I think the package owner should choose based on which they believe reflects where the larger portion of backing for the module comes from.

mcollina replied:
> In the case of a FOUNDATION it's harder to choose as it depends on the level of COMPANY support and at what level the package is a member of a foundation (for example IMPACT versus Sandbox). I think the package owner should choose based on which they believe reflects where the larger portion of backing for the module comes from.

I think we should not have this type of ambiguity in this format.

mcollina commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,115 @@
+* REGULAR-1 - There are dedicated resources who regularly maintain the module, expected response time is 1 days or less for
+  a "we read your issue" response. Further work will depend on prioritization of the issue by the maintainer team.
+* 24-7 - There are dedicated resources who regularly maintain the module and they are available 24/7. You can expect to 
+  be able to contact the maintainers and get an initial response with 6 hours.
```
This does not shows if a project has _paid_ 24-7 or REGULAR-1 support

rxmarbles commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,114 @@
+ 
+ ## Support-backing
+ 
+ * ABANDONED - There is nobody backing this module
```
I think we already would have `NONE`  as part of `support.target`. I do like the idea of a `HELP_WANTED` just not sure it would be under `Suppport Backing`

dominykas replied:
Just to make sure - did you mean to leave both `ABANDONED` and `NONE` in?

mhdawson commented:
I believe I have either updated or responded to all comments so far. Please take a look at let me know if there is something I missed or you have additional comments/suggestions before we should land this a draft.  If you think its good as a draft please add your review/approval.

dominykas commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,124 @@
+document considered "custom" and may irgnored for flagged by any tooling that
+consumes these elements in the package.json.
+
+In addiition a `url` field can optionally be provided with a link to more detailed
```
```suggestion
In addition a `url` field can optionally be provided with a link to more detailed
```

nschonni commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,124 @@
+    "target" : "LTS",
+    "response": "REGULAR-7",
+    "response-paid": "REGULAR-1",
+    "backing": "COMPANY"
```
Missing trailing comma

mhdawson replied:
fixed

mhdawson commented:
@wesleytodd you show as requesting changes. I think you comments have been addressed. Can you take another look?

@mcollina I think I have also addressed your comments or at least commented on questions.  Can you take another look as well.

mhdawson commented:
Sorry to nag but @mcollina @wesleytodd can you take a look?

mhdawson commented:
@mcollina addressed your comments please take another loo.

@we any chance you'll get to take a look soon?

Eomm commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,137 @@
+# Baseline practice - Document support support levels
```
I notice the file name is 
       Baseline pra**ctive** - Document support levels.md

Emuentes commented:
@Eomm has created a tool to validate the "support level" object in the package.json file.

Can some of the folks reviewing the spec review the code and provide any feedback if needed?

ghinks commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,137 @@
+  to the latest LTS Node.js version in order to ensure you can use new versions/get security fixes.
+* LTS - The package is maintained for the Node.js LTS releases (both in Active and Maintenence mode).
+  Anyone creating an application using an LTS version of Node.js and using the latest major version of
+  LTS adopting modules will will not have to accept semver-major level (ie. breaking) changes into that
```
will will - duplication of word

ghinks commented on docs/drafts/Baseline practive - Document support levels.md:
```diff
@@ -0,0 +1,137 @@
+identify your package with an option which is not yet listed, please PR your
+option into the lists in the sections which follow. In the package.json the values
+can be any string, those which are not documented in the lists within this
+document considered "custom" and may irgnored for flagged by any tooling that
```
typo ignored

mhdawson commented:
@ghinks , @mcollina , @Eomm comments addressed (the filename is fixed if you checkout the repo, Github UI does not seem to show it though)

mhdawson commented:
@wesleytodd see I got your mention wrong last time so tagging again.

wesleytodd commented:
Sorry, I have been attempting to disconnect and focus on my family for a bit.  I will be sure to check back in when my parental leave it finished, but for now I am going to continue my small hiatus.  I will approve so I don't hold anything up unnecessarily, and if I find any input I have when I check back in I will be sure to post it.  Keep up the good work team!

mhdawson commented:
I think I've addressed all the comments so far and since it's landing as a draft I think it's ok if it lands.  I'll plan to do that on Monday March 18th unless somebody comments/objects before then.
